### ISSUE TITLE
get_current_context not present in user_defined_macros

### ISSUE THREAD

**Issue by raphaelauv** [apache/airflow#33217]
### Apache Airflow version

2.6.3

### What happened

get_current_context() fail in a user_defined_macros

give
```
{abstractoperator.py:594} ERROR - Exception rendering Jinja template for task 'toot', field 'op_kwargs'. Template: {'arg': '{{ macro_run_id() }}'}
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/abstractoperator.py", line 586, in _do_render_template_fields
    rendered_content = self.render_template(
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/template/templater.py", line 168, in render_template
    return {k: self.render_template(v, context, jinja_env, oids) for k, v in value.items()}
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/template/templater.py", line 168, in <dictcomp>
    return {k: self.render_template(v, context, jinja_env, oids) for k, v in value.items()}
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/template/templater.py", line 156, in render_template
    return self._render(template, context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/abstractoperator.py", line 540, in _render
    return super()._render(template, context, dag=dag)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/template/templater.py", line 113, in _render
    return render_template_to_string(template, context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/helpers.py", line 288, in render_template_to_string
    return render_template(template, cast(MutableMapping[str, Any], context), native=False)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/helpers.py", line 283, in render_template
    return "".join(nodes)
  File "<template>", line 12, in root
  File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/sandbox.py", line 393, in call
    return __context.call(__obj, *args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/jinja2/runtime.py", line 298, in call
    return __obj(*args, **kwargs)
  File "/opt/airflow/dags/dags/exporter/finance_closing.py", line 7, in macro_run_id
    schedule_interval = get_current_context()["dag"].schedule_interval.replace("@", "")
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 784, in get_current_context
    raise AirflowException(
airflow.exceptions.AirflowException: Current context was requested but no context was found! Are you running within an airflow task?
```

### What you think should happen instead

User macros should be able to access to the current context 

```
airflow.exceptions.AirflowException: Current context was requested but no context was found! Are you running within an airflow task?
```

### How to reproduce

```python
from airflow.models import DAG
from airflow.operators.python import PythonOperator
from airflow.utils.dates import days_ago

def macro_run_id():
    from airflow.operators.python import get_current_context
    a = get_current_context()["dag"].schedule_interval.replace("@", "")
    if a == "None":
        a = "manual"
    return a


with DAG(dag_id="example2",
         start_date=days_ago(61),
         user_defined_macros={"macro_run_id": macro_run_id},
         schedule_interval="@monthly"):
    def toto(arg):
        print(arg)


    PythonOperator(task_id="toot", python_callable=toto,
                   op_kwargs={"arg": "{{ macro_run_id() }}"})

```

### Operating System

ubuntu 22.04

### Deployment

Other

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by ivan-afonichkin** [comment]
Hello @potiuk @raphaelauv, I would like to take this issue. Will submit PR in a few days. I'm not sure how I can assign it to myself though.

**Comment by eladkal** [comment]
> Hello @potiuk @raphaelauv, I would like to take this issue. Will submit PR in a few days. I'm not sure how I can assign it to myself though.

assigned

**Comment by josh-fell** [comment]
FWIW, this can be used as a workaround:
```py
from airflow.models import DAG
from airflow.operators.python import PythonOperator
from airflow.utils.dates import days_ago


def macro_run_id(dag):
    a = dag.schedule_interval.replace("@", "")
    if a == "None":
        a = "manual"
    return a


with DAG(
    dag_id="example2",
    start_date=days_ago(61),
    user_defined_macros={"macro_run_id": macro_run_id},
    schedule_interval="@monthly",
):

    def toto(arg):
        print(arg)

    PythonOperator(task_id="toot", python_callable=toto, op_kwargs={"arg": "{{ macro_run_id(dag) }}"})

```

**Comment by raphaelauv** [comment]
Thanks @josh-fell , it's a great hack for my (specific) use case 👍👍👍

**Comment by messi-gao** [comment]
@ivan-afonichkin I found that this issue was fixed at 2.7.1 version.is there a way to fix this issue under 2.7.1 version?

**Comment by potiuk** [comment]
> @ivan-afonichkin I found that this issue was fixed at 2.7.1 version.we may not upgrade airflow now , so is there a way to fix this issue under 2.7.1 version?

You can attempt to patch it manually - source of airflow is fully available and you might want to patch your source. But I heartily recommend to just upgrade airflow. Every single airflow release contains 10s of fixes to known problems in previous versions, a lot of them contains fixes to security issues and generally - almost by defintion - when you do not run latest version, you run a version that contains known bugs and security fixes. So "upgrade airflow to latest version" is abaolutely best way that you should proceed. 

You have to remember that even if it costs some time to upgrade, it will likely cost you (and your organisation) a lot more to NOT upgrade - because of known bugs that are fixed and potential risk of security issues you have not applied (and in the future when acts like CRA (Cyber Resilience Act ) in EU and similar in the US will be approved and become the law (it will be passed in EU in a month or two) your organisation will have to upgrade to latest security-fixed version of software anyway and it will cost penalties if they don't, so I suggest get used to situation that you upgrade software as soon as it is released.
